src/main/scala/lms/core/stub.scala

```diff
@@ -24,6 +24,16 @@ object Adapter extends FrontEnd {
   var typeMap: mutable.Map[lms.core.Backend.Exp, Manifest[_]] = _
   var funTable: List[(Backend.Exp, Any)] = _
 
+  def genGraph1(m1: Manifest[_], m2: Manifest[_])(prog: Exp => Exp) = {
+    genGraphCommon(m1, m2)(g.reify(prog))
+  }
+  def genGraphCommon(m1: Manifest[_], m2: Manifest[_])(prog: => Block) = {
+    typeMap = new scala.collection.mutable.HashMap[lms.core.Backend.Exp, Manifest[_]]()
+    funTable = Nil
+    var g: Graph = time("staging") { program(prog) }
+    (typeMap, g)
+  }
```

Comment:
Adding support to get graph (instead of directly codegen) so that we can transform the graph

Reply:
seems `g` doesn't have to be `var` here. Also, I was thinking if we really want to have a mutable `typeMap` in Adapter, it is confusing that when should we reset it or pass it to codegen.

Reply:
That is an interesting question. I think using `mutable typeMap` is probably faster than `immutable typeMap` when adding contents to the map? It is certainly temping to use `val` instead of `var` when the `typeMap` is mutable already. However, I am wondering if we want to reuse `Adapter` for transformation passes. If we do, we will have multiple graphs, each has a `typeMap` of its own. Then if `Adapter` is reused, we will need `Adapter.typeMap` to bind with a new hashMap. 

Or rather, just think about how one step of transformation can be done, when we want to use `Adapter.g.reflect` for the new (transformed) graph. We will have 2 typeMaps alive during the transformation. One is for the old graph, the other for the new graph. If both typeMaps are (or used to be) `Adapter.typeMap`, then I guess it has to be `var`?

Reply:
I think instead of populating the side effects of updating `typeMap` and `funTable`, maybe we should encapsulate them into `GraphBuilder` (or `Graph`) and manipulate them through the interface from `GraphBuilder` (or `Graph`). In the current status, there is a chance that `typeMap`, `funTable` and graph `g` are mismatched.

Reply:
So that the `typeMap` can be accessed via `Adapter.g.typeMap`.

Reply:
Not sure. Tiark might want to have the `Graph` type-free, because it represents the simple core of LMS-IR. The type is more like a Meta-language checker and a code-gen guidance.

With that said, if we can NOT change the Graph or GraphBuilder, but extend them with `TypedGraph` and `TypedGraphBuilder`, I guess Tiark will think that is OK.

Reply:
Having both untyped graph and TypedGraphsBuilder is okay to me, but if transformations can change the type, then it looks like we have to use a typed one.

Reply:
Even if transformation changes the type, it would generate a new typeMap I think. I don't think we `have to` use a typed graph, because a `typed graph` is "the same as" untyped graph + typeMap, right?